martha/martha_v3.js

```diff
@@ -468,22 +460,51 @@ async function marthaV3Handler(req, res) {
         `with auth required '${sendAuth}' and bond provider '${bondProvider}'`
     );
 
+    Object.assign(params, {
+        httpsMetadataUrl,
+        drsType,
+        sendAuth,
+        bondProvider,
+        accessMethodType,
+        bondSAKeyUrl,
+        bondAccessTokenUrl,
+    });
+}
+
+/**
+ * Retrieves information from the various underlying servers.
+ *
+ * See also:
+ * - https://bvdp-saturn-dev.appspot.com/#workspaces/general-dev-billing-account/DRS%20and%20Signed%20URL%20Development%20-%20Dev
+ * - https://lucid.app/lucidchart/invitations/accept/0f899643-76a9-4b9c-84f5-f11ddac86bba
+ * - https://lucid.app/lucidchart/invitations/accept/8b6f942b-f7dc-4acc-ac36-318a1685e6ac
+ */
+async function retrieveFromServers(params) {
+    const {
+        requestedFields,
+        httpsMetadataUrl,
+        auth,
+        drsType,
+        sendAuth,
+        accessMethodType,
+        bondSAKeyUrl,
+        bondAccessTokenUrl,
+    } = params;
+
     let response;
     if (overlapFields(requestedFields, MARTHA_V3_METADATA_FIELDS)) {
         try {
             response = await apiAdapter.getJsonFrom(httpsMetadataUrl, sendAuth ? auth : null);
         } catch (error) {
-            logAndSendServerError(res, error, 'Received error while resolving DRS URL.');
-            return;
+            throw new RemoteServerError(error, 'Received error while resolving DRS URL.');
         }
     }
 
     let drsResponse;
     try {
         drsResponse = responseParser(response);
```

Comment:
Can this even throw an error? I didn't notice anything in `responseParser`.

Reply:
Nevermind. There's code in `responseParser` that deals with converting DOS responses to DRS responses that I suspect could throw errors because I don't have enough knowledge of the DOS spec. I retract my implied suggestion to remove the `try`/`catch` from here.